plugins/dell-dock2/fu-dell-dock2-ec-hid.c

```diff
@@ -0,0 +1,240 @@
+{
+	guint8 *outbuffer = (guint8 *)user_data;
+	return fu_hid_device_set_report(FU_HID_DEVICE(self),
+					0x0,
+					outbuffer,
+					192,
+					DELL_DOCK2_EC_HID_TIMEOUT * 3,
+					FU_HID_DEVICE_FLAG_NONE,
+					error);
+}
+
+static gboolean
+fu_dell_dock2_hid_set_report(FuDevice *self, guint8 *outbuffer, GError **error)
+{
+	return fu_device_retry(self,
+			       fu_dell_dock2_hid_set_report_cb,
+			       DELL_DOCK2_EC_HID_MAX_RETRIES,
+			       outbuffer,
+			       error);
+}
+
+static gboolean
+fu_dell_dock2_hid_get_report_cb(FuDevice *self, gpointer user_data, GError **error)
+{
+	guint8 *inbuffer = (guint8 *)user_data;
+	return fu_hid_device_get_report(FU_HID_DEVICE(self),
+					0x0,
+					inbuffer,
+					192,
+					DELL_DOCK2_EC_HID_TIMEOUT,
+					FU_HID_DEVICE_FLAG_NONE,
+					error);
+}
+
+static gboolean
+fu_dell_dock2_hid_get_report(FuDevice *self, guint8 *inbuffer, GError **error)
+{
+	return fu_device_retry(self,
+			       fu_dell_dock2_hid_get_report_cb,
+			       DELL_DOCK2_EC_HID_MAX_RETRIES,
+			       inbuffer,
+			       error);
+}
+
+gboolean
+fu_dell_dock2_hid_raise_mcu_clock(FuDevice *self, gboolean enable, GError **error)
+{
+	FuEcHIDCmdBuffer cmd_buffer = {
+	    .cmd = HUB_CMD_WRITE_DATA,
+	    .ext = HUB_EXT_MCUMODIFYCLOCK,
+	    .cmd_data0 = (guint8)enable,
+	    .cmd_data1 = 0,
+	    .cmd_data2 = 0,
+	    .cmd_data3 = 0,
+	    .bufferlen = 0,
+	    .parameters = {.i2ctargetaddr = 0, .regaddrlen = 0, .i2cspeed = 0},
+	    .extended_cmdarea[0 ... 52] = 0,
+	};
+
+	if (!fu_dell_dock2_hid_set_report(self, (guint8 *)&cmd_buffer, error)) {
+		g_prefix_error(error, "failed to set mcu clock to %d: ", enable);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dell_dock2_hid_erase_bank(FuDevice *self, guint8 idx, GError **error)
+{
+	FuEcHIDCmdBuffer cmd_buffer = {
+	    .cmd = HUB_CMD_WRITE_DATA,
+	    .ext = HUB_EXT_ERASEBANK,
+	    .cmd_data0 = 0,
+	    .cmd_data1 = idx,
+	    .cmd_data2 = 0,
+	    .cmd_data3 = 0,
+	    .bufferlen = 0,
+	    .parameters = {.i2ctargetaddr = 0, .regaddrlen = 0, .i2cspeed = 0},
+	    .extended_cmdarea[0 ... 52] = 0,
+	};
+
+	if (!fu_dell_dock2_hid_set_report(self, (guint8 *)&cmd_buffer, error)) {
+		g_prefix_error(error, "failed to erase bank: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_dell_dock2_hid_i2c_write(FuDevice *self,
+			    const guint8 *input,
+			    gsize write_size,
+			    const FuHIDI2CParameters *parameters,
+			    GError **error)
+{
+	FuEcHIDCmdBuffer cmd_buffer = {
+	    .cmd = HUB_CMD_WRITE_DATA,
+	    .ext = HUB_EXT_I2C_WRITE,
+	    .dwregaddr = 0,
+	    .bufferlen = GUINT16_TO_LE(write_size),
+	    .parameters = {.i2ctargetaddr = parameters->i2ctargetaddr,
+			   .regaddrlen = 0,
+			   .i2cspeed = parameters->i2cspeed | 0x80},
+	    .extended_cmdarea[0 ... 52] = 0,
+	};
+
+	g_return_val_if_fail(write_size <= HIDI2C_MAX_WRITE, FALSE);
+
+	memcpy(cmd_buffer.data, input, write_size);
```

Comment:
fu_memcpy_safe

Reply:
Actually, I think all of this is screaming out for a conversion to rustgen -- I might take a stab at that this afternoon.

Reply:
How about check-source.py catches any calls of `memcpy(` and makes noise?

Reply:
Agree, and also maybe `__attribute__((packed))` too?

Reply:
Ah good idea, that should mean rust gen.

Reply:
Fixed, thanks.